inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp

```diff
@@ -737,6 +751,80 @@ void InsertCopyLayerPass::run() {
     }
 }
 
+void Concat4Dto2DPass::run() {
+    auto quantized = InferenceEngine::getInjectedData<QuantizedLayerParams>(pLayers->front());
+
+    if (getPassManager()->getPolicy().ConcatConversionPolicy == Policy::Concat4Dto2DConversion::DISABLED) return;
+    if (getPassManager()->getPolicy().ConcatAlignmentPolicy == Policy::ConcatAlignment::DISABLED) return;
+    if (getPassManager()->getPolicy().ConcatAlignmentPolicy == Policy::ConcatAlignment::DISABLED_FOR_FP32 && !quantized) return;
+
+    for (auto & l : *pLayers) {
+        LayerInfo info(l);
+        if (!info.isConcat()) continue;
+        auto concatLayer = info.as<ConcatLayer*>();
+
+        auto dims_size = concatLayer->insData[0].lock()->getDims().size();
+        if (dims_size > 2) {
+            auto axis = concatLayer->_axis;
+            bool skip_layer = false;
+            for (int i = 0; i < axis; i++) {
+                if (concatLayer->insData[0].lock()->getDims()[i] != 1) skip_layer = true;
+            }
+            if (skip_layer) continue;
+            skip_layer = true;
+            for (auto input : concatLayer->insData) {
+                auto input_dims = input.lock()->getDims();
+                auto total_size = std::accumulate(input_dims.begin(), input_dims.end(), static_cast<size_t>(1), std::multiplies<size_t>());
+                if (total_size % 64 != 0) skip_layer = false;
+            }
+            if (skip_layer) continue;
+
+            for (auto input_idx = 0; input_idx != concatLayer->insData.size(); input_idx++) {
+                auto getLayerByIndex = [&concatLayer](int idx) {
+                    auto input = concatLayer->insData[idx];
+                    auto lockedInput = input.lock();
+                    if (!lockedInput) {
+                        THROW_GNA_EXCEPTION << "cannot get insdata : "<< idx << " for layer: " << concatLayer->name;
+                    }
+                    return lockedInput;
+                };
+
+                auto concatInput = getLayerByIndex(input_idx);
+                auto input_dims = concatInput->getDims();
+                auto total_size = std::accumulate(input_dims.begin(), input_dims.end(), static_cast<size_t>(1), std::multiplies<size_t>());
```

Comment:
It seems like the same operation is done twice. here and above in the skip layer check

Reply:
it is, but i wanted to avoid creating some vector to store the result of first calculation.